pyo3-polars-derive/src/lib.rs

```diff
@@ -12,14 +12,21 @@ fn create_expression_function(ast: syn::ItemFn) -> proc_macro2::TokenStream {
         use pyo3_polars::export::*;
         // create the outer public function
         #[no_mangle]
-        pub unsafe extern "C" fn #fn_name (e: *mut polars_ffi::SeriesExport, len: usize) -> polars_ffi::SeriesExport {
+        pub unsafe extern "C" fn #fn_name (e: *mut polars_ffi::SeriesExport, len: usize, kwargs: *const std::os::raw::c_char) -> polars_ffi::SeriesExport {
```

Comment:
As in, the intent is to json- (or whatever-else-) serialize it on the python side and unwrap it back here, or something along the lines? If so, that's a bit cheesy for sure 😆

Reply:
Yes, that's the idea. And then users can work the `JsonValue`.

Any idea of something more ergonomic? :eyes:?

Reply:
I think with floats in particular that's a problem waiting to happen; and with polars being a numeric library that may be VERY undesirable...

Like use pass X.000000000 and get Y.99999999999 due to serialization/deserialization and then spend days debugging it. In some cases that may not matter, but in some cases it may change the output.

If we go this way (i.e. serialization/deserialization using some standard protocol), then why do we have to use JSON at all? Can we use any binary serde-compatible serialization protocol, that can pack args from within Python? (and btw does it have to be done in pure Python, or in Rust in py-polars?) Something like msgpack? Need to brush up on the formats.

---

Another thought... why not allow the 'receiver' side (that is, the user Rust code for the plugin) to simply serde-deserialize the kwarg struct from, say, raw bytes instead of poking manually at those fields and types? This would also alleviate the need for manual error handling during kwargs deserialization etc, and would make it feel more statically typed and rusty overall.

(The only question is then how to serialize those args on the 'sender' side, where exactly does the serialization take place and which protocol to use).

I.e. wouldn't it be nice if you could
```rust
// 'receiver' side
#[derive(Deserialize)]
struct Kwargs<'a> {
   float_arg: f64,
   string_arg: &'a str,  
}

let args = serde_something::from_bytes(&args_payload)?;
```

Reply:
Save from the lifetime on the kwargs string, this worked like a charm! Thanks for keeping me honest here ;)

Reply:
👍 @ritchie46 Looks great! Re: ergonomics, wonder if `Option<>` has to always be there around kwargs? Because.. if it's `DefaultKwargs` then an empty dict is valid kwargs anyways, just like in Python. And if you require your own kwargs you can always add an `Option<>` yourself if needed (to either the whole type or some/all of the fields)?

Reply:
Yes, we can do that as well. Need to do more proc-macro inspection for that.